### PR TITLE
Refactor template builder

### PR DESCRIPTION
**Pull request by evgenyz**
Unify the code and remove duplication.

Introduce a basic test for template builder.

Based on changes from #9833.

### CODE REVIEW

**Review by jan-cerny — COMMENTED** [review]
I think that it would be better if each of the commits would exist as separate PRs.

File: tests/unit/ssg-module/test_templates.py
Comment (evgenyz) [reply tests/unit/ssg-module/test_templates.py:32]: I'm sure that @matejak would not agree with you on this.

But regardless it is just an introduction of tests for template building, they will evolve in consecutive PRs.

**Comment by jan-cerny** [comment]
I was confused that these are 2 PRs, but now it's clear to me. We usually write a message for example: This PR is blocked by #XYZ and will be rebased on top after will be rebased on top after it will be merged

**Comment by evgenyz** [comment]
> I was confused that these are 2 PRs, but now it's clear to me. We usually write a message for example: This PR is blocked by #XYZ and will be rebased on top after will be rebased on top after it will be merged

Sorry about the confusion. Rebased.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:253]: This will lead to dragging rule_id and rule_title even further down the pipe. Ideally I would like to organize this context in a more opaque and organized way, but it'll come later.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:42]: Wouldn't it be lovely to call this a `template_root`? While template path is not wrong, I think that root is better, as a template is a collection of files rather than one thing where you can point with a path and that's it.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:40]: This is a wrong name, it's a root directory of template**s**, not of a template.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:74]: The comment is not needed, or you can make it literal by writing
```
if self.preprocessing_file_path is None:
    return None
other stuff 1
other stuff 2
```

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:135]: While you are at refactoring, would it be possible to extract mapping of output directories and mapping of templates into private methods? If you give them correct names, you don't need comments any more.

File: tests/unit/ssg-module/test_templates.py
Comment (matejak) [reply tests/unit/ssg-module/test_templates.py:32]: Indeed, you don't know how the output should *exactly* look like, you don't want to write tests that are too specific, because those tests would be fragile. As the template code can be subject of change, you want to test only selected general assumptions to make sure that the ends meet.
This is different from testing XCCDF output, as there is only one right way how to represent certain data in the XCCDF format, and in such case, the test comparing the generated XML against a reference isn't fragile.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:293]: This comment has become obsolete.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:234]: Let's get rid of the name `extra_env`, because on this level of abstraction, we don't care about origins of the dictionary. Isn't it just a dictionary with keys and substitutions?

**Comment by matejak** [comment]
It would generally be nice to have a PR description that would introduce submitted changes, and somehow explain their benefits, and even their tradeoffs that were introduced or exposed, but not addressed. Right now, reviewers will focus on low-level details of the changed code, and those design changes may not be understood correctly.

File: ssg/templates.py
Comment (jan-cerny) [reply ssg/templates.py:293]: @matejak No, it hasn't. We still pass `oval_def_id` instead of `rule_id` to the `self.build_lang` on line 294.

File: ssg/templates.py
Comment (matejak) [inline ssg/templates.py:143]: What was the motive of moving from rule properties to concrete fields?

- On the good side, arguments of the method got narrowed down more, which is always good, but
-  the method itself references rules in in messages and in its docstring.

The method is used in rule context only, so shouldn't the new arguments be renamed to `rule_id` and `rule_template` then?

Comment (evgenyz) [reply ssg/templates.py:143]: Everything is even simpler: the list of languages is based on the properties of a template and template only. Even the id is used only to show an error message (very controversial decision, but I left it there for to minimize the amount of changes to review). So, the context of the method is template. And even the method itself could be static or a module function.

Comment (evgenyz) [reply ssg/templates.py:143]: Scratch that. After giving it another 15 seconds I realized that even the message is bogus. If the function does not depend on the properties of a rule there is no sense in mentioning its ID.

File: ssg/templates.py
Comment (matejak) [reply ssg/templates.py:253]: I think that I agree with both of you, so I would do it the other way around - Evgeny is correct that the `render_lang_file` doesn't know anything specific about rules, and it just has that dictionary with everything that it can use without going into details.
And Jan is correct as well - so why not to move the addition of the product from the "env yaml" one level above, so modifications of this dictionary is localized in one method, and the rendering method becomes more abstract?

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:143]: Err. Actually no, it's a bit more complicated. The template is not the template but a subsection of rule settings. Oh, well.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:253]: It can't be done ATM because of https://github.com/ComplianceAsCode/content/blob/9c9f7c2ad405139bd5bc009dfaf3bb443e21b148/ssg/build_sce.py#L169

File: tests/unit/ssg-module/test_templates.py
Comment (evgenyz) [reply tests/unit/ssg-module/test_templates.py:32]: I'd simply say that in a unit test one should focus on testing things that are coming via function arguments and avoid testing side-effects at all costs.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:42]: In these names there is usually a dichotomy of `file_name` and `path`. The name `template_root` does not immediately identify itself as a file name or a path and gives a false impression that there are sensible locations deeper inside this root path.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:40]: Yeah.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:74]: Sure.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:135]: NP.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:266]: Yep.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:236]: That's a drawback of merging a feature step-by-step. You don't see full picture. It will be used for platforms in a different way later.

File: tests/unit/ssg-module/test_templates.py
Comment (evgenyz) [reply tests/unit/ssg-module/test_templates.py:32]: So there shall be other tests that would catch proper ID, product, etc. This particular test at this moment ensures that the title is propagated.

File: ssg/templates.py
Comment (jan-cerny) [inline ssg/templates.py:152]: The Code Climate now started to complain about a cognitive complexity of this method. You can probably try to extract a method from the code in the `if` block.

Comment (evgenyz) [reply ssg/templates.py:152]: Made it simpler, let's see what CC would say about this attempt.

Comment (evgenyz) [reply ssg/templates.py:152]: So, it's fixed in eyes of CC.

**Comment by evgenyz** [comment]
I don't think that remaining CC warnings are important.

**Review by jan-cerny — APPROVED** [review]
thanks for all the improvements

**Review by matejak — CHANGES_REQUESTED** [review]
Looks good, but there is one thing that needs some work, because as it is, it is a source of confusion. See my comment near the `build_extra_ovals` method.

File: ssg/templates.py
Comment (evgenyz) [reply ssg/templates.py:241]: These variables, `rule_id` and `rule_title` are hardcoded in tons of templates and macros, and templates are expecting them even if they are rendered for "extra ovals" or even platforms. And it kinda makes sense to not have it named `check_id` or `check_title` because this base value is used not only in checks but in remediations as well (and would be in platforms).

Passing `rule_id` into functions that supposed to be agnostic to this level of abstractions is *very ugly*, but it makes perfect sense in practice: whenever a dev screws a build by changing something in a rule it won't help knowing what template caused it. Other way around is pretty straightforward — every rule has just one template.

Also, platforms would add another slice of complexity here because of second layer of indirection.

The whole template builder code (and rendering) is written upside-down, and IMHO should be rewritten (including Jinja part), but it is a huge scope creep.

One thing I could promise to you right now though: there would be one less extra_env=None in the next PR.

Comment (matejak) [reply ssg/templates.py:241]: Let's forget about introduction of `check_id`, or about renaming `rule_id` to `id` and so on - we both agree that those things are outside of the scope of this PR. However, if something is named `rule_id` and I have to pass it an ID that isn't a rule ID to get what I need, then the code should be refactored, because it confuses its users.
Let's also forget about "extra_env"'s name. On that level of abstraction, it is "stuff for the template that the method should to pass on, and perhaps add some extra stuff there", but we rename it later.

Usage of `rule_id` in the code that returns the template name is slightly unfortunate, because throwing an exception would allow to provide the context on a higher level, but I also don't suggest to change this, as this is not the code touched by this PR.

I suggest to introduce this generic function that would sit between `build_rule`/`build_extra_ovals` and `build_lang` - the `render_lang_file` that you propose is too low-level to be conveniently used directly. In order to do that easily, I propose to move rule ID and rule title to that "extra env" dictionary. You will still need to put something for `get_template_name`, but "extra OVAL ID ..." will do just right for the time being.

Comment (evgenyz) [reply ssg/templates.py:241]: Okay, I'm looking at a newer code already. So there is a bit of disconnect. In the future that you're not seeing yet there is no `build_lang`: `build_rule` -> `get_lang_contents` -> `render_lang_file`. That's it. The function `get_lang_contents` is the mediator (or would be). But I really don't want to bring another couple dozen lines in this PR. We will have an opportunity to discuss it once more with platform templates rendering present in the code.

Comment (matejak) [reply ssg/templates.py:241]: Great, yes, you are correct, you will have to repeat it, but that's fine. You compose "the stuff for the template", and it is expected that you build that with slightly different intentions in each of those cases. It may look weird, as the "env yaml" keys will still be rule-centric, but we can discuss that later. Right now, it is important that the code allows different way of filling that dictionary.

Regarding `get_lang_contents` - I suggest slightly changing its API to something like
```
get_lang_contents(self, rule_id_context, template, language, extra_env)
```
and remove statements that populate `extra_env` in lower-level functions.

`build_sce.checks` is an interesting one - you actually have the "stuff for the template" at your disposal it just waits to be passed on, so you can even make the somehow simpler: https://github.com/ComplianceAsCode/content/blob/master/ssg/build_sce.py#L128

Comment (matejak) [reply ssg/templates.py:241]: This is a refactoring PR, and it is not done, because the code that has been touched has a significant design smell. Therefore, I won't merge this PR until the smell is dealt with, we don't use future as an argument when reviewing PRs.

Comment (evgenyz) [reply ssg/templates.py:241]: Python is not a strongly typed language, but there is a reason we don't just pass around all arguments in a generic dictionary.

Comment (evgenyz) [reply ssg/templates.py:241]: If you insist that implicit required arguments for templates named `rule_id` and `rule_title` is an important problem, and it should be dealt with right away — I shall try and do that.

Comment (jan-cerny) [reply ssg/templates.py:241]: OK So do I understand it well that there won't be any further changes in this PR?

Comment (jan-cerny) [reply ssg/templates.py:241]: I'd prefer the first option. Using a dictionary for passing parameters could make the function less readable because people need to make more effort to discover the actual contents of the dictionary.

Comment (matejak) [reply ssg/templates.py:241]: That dictionary is a list of substitutions, its "customer" is a template, not lower-level functions down the road. So having stuff there doesn't affect the readability, because you don't have reasons to inspect it, it is just a package of static data.

Concerning any contract-related stuff, the only proper way of handling this is to make Jinja more explicit about undefined values, which BTW can be done. If we go for something weaker s.a. assuring that those keys are available to the templates, a more solid way than supplying arguments to higher-level functions would be to check the dictionary further down for keys, e.g. in `render_lang_file`. So I don't see a point here either.

My objection related to the current PR isn't template field names, because those have already been here before the PR. The problem is that the refactored code has confusing design, which can be demonstrated in the relationship between `build_extra_ovals` and `build_lang`. It may look like a small thing, but it isn't, because it creates an implicit assumption, it makes the code confusing, and we refactor exactly in order to avoid such things.

**Comment by evgenyz** [comment]
This PR is superseded by #9870.

**Pull request closed without merging**
